rlbottraining/common_graders/rl_graders.py

```diff
@@ -29,13 +30,77 @@ def __init__(self, timeout_seconds=4.0, ally_team=0):
         ])
 
 class FailOnBallOnGroundAfterTimeout(FailOnTimeout):
+    def __init__(self, max_duration_seconds):
+        super().__init__(max_duration_seconds)
+        self.previous_ang_x = None
+        self.previous_ang_y = None
+        self.previous_ang_z = None
+        self.previous_total_goals = None
+
+    def set_previous_angular_velocity(self, ball, reset = False):
+        if not reset:
+            self.previous_ang_x = ball.angular_velocity.x
+            self.previous_ang_y = ball.angular_velocity.y
+            self.previous_ang_z = ball.angular_velocity.z
+            self.previous_vel_z = ball.velocity.z
+        else:
+            self.previous_ang_x = None
+            self.previous_ang_y = None
+            self.previous_ang_z = None
+            self.previous_vel_z = None
+
+    def set_previous_total_goals(self, total_goals, reset=False):
+        if not reset:
+            self.previous_total_goals = total_goals
+        else:
+            self.previous_total_goals = None
+
+    def current_total_goals(self, packet):
+        total_goals = 0
+        for car_id in range(packet.num_cars):
+            goal = packet.game_cars[car_id].score_info.goals
+            own_goal = packet.game_cars[car_id].score_info.own_goals
+            total_goals += goal + own_goal
+        return total_goals
 
     def on_tick(self, tick: TrainingTickPacket) -> Optional[Grade]:
         grade = super().on_tick(tick)
         if grade is None:
             return None
         assert isinstance(grade, FailOnTimeout.FailDueToTimeout)
         ball = tick.game_tick_packet.game_ball.physics
-        if ball.location.z < 100 and ball.velocity.z >= 0:
-            return grade
+        hit_ground = False
+
+        if self.previous_ang_z is None:
+            self.set_previous_angular_velocity(ball)
+        else:
+            max_ang_vel = 5.9999601985025075  #Max angular velocity possible
+            previous_ang_norm = math.sqrt(self.previous_ang_x**2 + self.previous_ang_y**2 + self.previous_ang_z**2)
+            if ball.location.z <= 1900:  #Making sure it doesnt count the ceiling
+                if (ball.angular_velocity.x != self.previous_ang_x or ball.angular_velocity.y != self.previous_ang_y):
+                    # If the ball hit anything its angular velocity will change in the x or y axis
+                    if self.previous_ang_z == ball.angular_velocity.z and not ( 130 < ball.location.z < 150) :
+                        # if the z angular velocity did not change it means it was a flat plane.
+                        #ignore dribbling
+                        hit_ground = True
+                    elif previous_ang_norm == max_ang_vel:
```

Comment:
I'd recommend not doing exact equality on floating point numbers.
Do we even have cases that get into this? 
Could you please add unit tests?

Reply:
I tested various times and the max was always like that. I would think that it could have some error from time to time, but all the tested times had exactly that value.
Yes, I believe it is shot 7 on the linkuru pack, it will have max angular velocity when hitting the ground.

Reply:
I would like some help with how should I add unit testing, I'm not familiar with how you are doing that, neither how that is done on general

Reply:
https://github.com/RLBot/RLBotPythonExample/blob/master/training/unit_tests.py

Tests go here: https://github.com/RLBot/RLBotTraining/tree/master/tests
To run all tests, run `test_all.bat` in the root of this repo. `test_quick.bat` is for anything that completes in less than a second (i.e. does not launch rocket league)
To run tests only that file, do the usual `python test_my_foo.py`

To test these kinds of cases, create exercises within the test file itself which use this grader. e.g.
- an exercise with zero seconds remaining and the ball dropping onto the car from a height should fail after a specific time (which is longer than what it takes to bounce off the car once).
- an exercise with zero seconds remaining where the ball drops onto specific locations of the goal and it should quickly end in within a small period of seconds.

Reply:
Ah, thank you, this will also make my testing easier :)
I will try to implement this this afternoon